Feature/anchor buttons

paulmsmith opened the pull request:
As discussed in '[remove button role from start link](https://github.com/alphagov/govuk_elements/pull/272)' here is a pull request to shim anchors with ```role="button"``` to be 'clicked' via the space key.

I've not written tests yet, assuming the general style of this is fine then I'll do that. My thoughts were that this shim should be configurable but I've heard opinion that for a 'shim' that is not always a good thing. Thoughts, Suggestions welcome. :)

alextea commented on javascripts/govuk/anchor-buttons.js:
```diff
@@ -0,0 +1,55 @@
+
+  };
+
+  // if disbaleInit is not true then run the init method
```
typo ;)

paulmsmith replied:
Good spot! Will fix.

nickcolley commented on javascripts/govuk/anchor-buttons.js:
```diff
@@ -0,0 +1,55 @@
+    
+    // init function (so it can be executed again if needed)
+    init: function init() { 
+      var $elms = $(this.config.selector);
```
Could we attach a listener to the document rather than one for each button?

```js
$(document).on('keyup', this.config.selector, this.triggerClickOnTarget.bind(this));
```

My thought is that this way we only have one delegated listener rather than potentially a few?

paulmsmith replied:
Yep, good point I think that's probably better. 👍  Will do that.

nickcolley commented on javascripts/govuk/anchor-buttons.js:
```diff
@@ -0,0 +1,55 @@
+      // setting to true will disable the init function upon execution
+      disableInit: false,
+      // the target element(s) to attach the shim event to
+      selector: 'a[role="button"]',
```
Can we add this role to other elements such as spans?

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_button_role

Maybe the selector should be just `[role="button"]`.

paulmsmith replied:
I wasn't sure about that one. Do we always want space to trigger a click on something with ```role="button"``` ? @LJWatson @aduggin to advise perhaps?

nickcolley commented:
Thanks a lot for getting this together Paul 👍 

One thing that worries me with this is it could break the behaviour of scrolling the page when pressing space.

It seems as though if a user is focused on a link it's okay to block that behaviour but I'm not entirely sure of all the edge cases so thought I'd comment in the hope that someone else can think about this too.

nickcolley commented on javascripts/govuk/anchor-buttons.js:
```diff
@@ -0,0 +1,55 @@
+    triggerClickOnTarget: function triggerClickOnTarget(event) {
+      var code = event.charCode || event.keyCode; 
+      // if the keyCode/charCode from this event is in the keycodes array then 
+      if ($.inArray(code, this.config.keycodes) !== -1) {
```
A nit (sorry) but I'm wondering, if we're only ever going to be shimming the spacekey, could we just do a boring

```js
code === 32
```

here instead of an array comparison, not super important (sorry again :P)

paulmsmith replied:
Again fair point. Originally there was talk of triggering on enter as well. Left it open to additional key-codes just in case more than anything. Happy to change if we feel that makes it more readable/explicit.

nickcolley replied:
Generally for me I prefer to avoid anticipating future updates since more often than not they don't happen but don't think this blocks merging, so up to you. ✌️

robinwhittleton commented on javascripts/govuk/anchor-buttons.js:
```diff
@@ -0,0 +1,55 @@
+(function(global, customConfig) {
```
Can we add a comment block at the top explaining what the file does? E.g. https://github.com/alphagov/govuk_frontend_toolkit/blob/master/javascripts/govuk/stop-scrolling-at-footer.js

robinwhittleton commented on javascripts/govuk/anchor-buttons.js:
```diff
@@ -0,0 +1,55 @@
+(function(global, settings) {
```
Can we add a comment block at the top explaining what the file does? E.g. https://github.com/alphagov/govuk_frontend_toolkit/blob/master/javascripts/govuk/stop-scrolling-at-footer.js

paulmsmith replied:
Sure..  I'll get that done and reply back when I've updated this PR.

robinwhittleton commented:
This adds a file, but won’t roll it out across GOV.UK. I guess we want to add it to [govuk_template.js](https://github.com/alphagov/govuk_template/blob/master/source/assets/javascripts/govuk-template.js), intialise it in [govuk_template’s core.js](https://github.com/alphagov/govuk_template/blob/master/source/assets/javascripts/core.js), then push a new version and update [static’s Gemfile](https://github.com/alphagov/static/blob/master/Gemfile) to use that.

paulmsmith commented:
Hi all, I've made the updates after the suggestions. I have changed it now so that the event listener is on the document and I have also updated the module pattern so that it needs to be explicitly initiated.

paulmsmith commented:
@robinwhittleton I'm due to go away for two weeks now. I've removed the 'WIP' from the title and I think this can be merged whenever you and others feel is a good time.

gemmaleigh commented:
@paulmsmith thanks for this! We'll review shortly. Have a good break!

gemmaleigh commented:
I'd like to get @aduggin's opinion on this before we merge.

gemmaleigh commented on javascripts/govuk/anchor-buttons.js:
```diff
@@ -0,0 +1,61 @@
+    },
+
+    // By default this will find all anchors with role attribute set to 
+    // 'button' and will trigger their click event when the spaceky (32) is pressed.
```
mini-nit, typo in comment (spacekey)

gemmaleigh commented on javascripts/govuk/anchor-buttons.js:
```diff
@@ -0,0 +1,61 @@
+  // hand back to global
+  global.GOVUK = GOVUK;
+      
+})(window);
```
Missing EOF.

aduggin commented:
This approach looks good to me

robinwhittleton commented:
Thanks. Will merge this now, address @gemmaleigh’s nits in a further PR.
